hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java

```diff
@@ -0,0 +1,263 @@
+            } catch (Exception e) {
+                logger.severe("Error completing publication for publisher " + publisher, e);
+            }
+        }
+    }
+
+    public LiveOperationRegistry getLiveOperationRegistry() {
+        return liveOperationRegistry;
+    }
+
+    @Override
+    public void populate(LiveOperations liveOperations) {
+        liveOperationRegistry.populate(liveOperations);
+    }
+
+    /**
+     * Read metrics from the journal from the given sequence.
+     *
+     * @param startSequence The sequence start reading the metrics with.
+     */
+    public CompletableFuture<RingbufferSlice<Map.Entry<Long, byte[]>>> readMetrics(long startSequence) {
+        if (!config.isEnabled()) {
+            throw new IllegalArgumentException("Metrics collection is not enabled");
+        }
+        CompletableFuture<RingbufferSlice<Map.Entry<Long, byte[]>>> future = new CompletableFuture<>();
+        future.whenComplete(withTryCatch(logger, (s, e) -> pendingReads.remove(future)));
+        pendingReads.put(future, startSequence);
+
+        tryCompleteRead(future, startSequence);
+
+        return future;
+    }
+
+    private void tryCompleteRead(CompletableFuture<RingbufferSlice<Map.Entry<Long, byte[]>>> future, long sequence) {
+        try {
+            RingbufferSlice<Map.Entry<Long, byte[]>> slice = metricsJournal.copyFrom(sequence);
+            if (!slice.isEmpty()) {
+                future.complete(slice);
+            }
+        } catch (Exception e) {
+            logger.severe("Error reading from metrics journal, sequence: " + sequence, e);
+            future.completeExceptionally(e);
+        }
+    }
+
+    @Override
+    public void reset() {
+    }
+
+    @Override
+    public void shutdown(boolean terminate) {
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel(false);
+        }
+
+        for (MetricsPublisher publisher : publishers) {
+            try {
+                publisher.shutdown();
+            } catch (Exception e) {
+                logger.warning("Error shutting down metrics publisher " + publisher.name(), e);
+            }
+        }
+    }
+
+    private List<MetricsPublisher> getPublishers() {
+        List<MetricsPublisher> publishers = new ArrayList<>();
+        if (config.isEnabled()) {
+            int journalSize = Math.max(
+                    1, (int) Math.ceil((double) config.getRetentionSeconds() / config.getCollectionIntervalSeconds())
+            );
+            metricsJournal = new ConcurrentArrayRingbuffer<>(journalSize);
+            ManagementCenterPublisher publisher = new ManagementCenterPublisher(this.nodeEngine.getLoggingService(),
+                    (blob, ts) -> {
+                        metricsJournal.add(entry(ts, blob));
+                        pendingReads.forEach(this::tryCompleteRead);
+                    }
+            );
+            publishers.add(publisher);
+        }
+
+        if (config.isJmxEnabled()) {
```

Comment:
we also changed the logic here, so the `isEnabled` is like a master switch, I don't know how you want to make the configuration. so if isEnabled is false, you can't enable just JMX.

Reply:
That means you can't disable publishing to MC (buffers) if you want JMX only. How if we have this in the config?
```
public boolean isMetricsEnabled() {
  return isMcEnabled() || isJmxEnabled();
}
```

And we could still switch MC or JMX independently.

Reply:
How about adding the enabled/disabled information accessible via MetricsService ? This way we don't need to go and fetch the Config object from the NodeEngine and access MetricsConfig from there. We already have MetricsService and it has the MetricsConfig.

Reply:
2 cents from me. DS stats are enabled with the following check:
https://github.com/hazelcast/hazelcast/pull/15560/files#diff-b3e262104ca2692463bc0f125cb5fd22R217

If I'm getting everything right, it means that users won't be able to enable JMX and DS stats only, without enabling MC metrics. If it's correct, then it looks a bit weird to me.

Reply:
Yes, that's indeed better than doing it in the config.

Reply:
@puzpuzpuz this is a valid point, especially if we consider registering custom publishers. I think 
- enabling DS metrics should be independent from enabling MC or JMX publishers
- enabling MC and JMX publishers should also be independent from each other

Reply:
Well MC collection doesn't do anything by itself except to store the latest blob. If metrics will be rendered anyway I think you can always enable MC collection?

Reply:
> How about adding an access to enabled/disabled information via MetricsService ? This way we don't need to go and fetch the Config object from the NodeEngine and access MetricsConfig from there. We already have MetricsService and it has the MetricsConfig.

Well you pass a function to register publisher and if metrics are disabled, the function shouldn't be called?

Reply:
> If metrics will be rendered anyway I think you can always enable MC collection?

In this case, rendering and storing those blobs will consume some CPU time and memory even if blobs aren't used at all. I'd prefer to pay for it only when MC metrics collection is enabled.

Reply:
Finally, introduced a master switch along with `mc-enabled` flag, so populating metrics to the ringbuffer can be turned off. The master switch is needed because we support programmatic publisher registration that is expected to ignore the registration if the metrics system is entirely switched off. It is a valid case that both MC and JMX publishers are disabled, but we want to allow programmatic registration of publishers. It can't be done without a master switch.